frame/support/src/storage/bounded_vec.rs

```diff
@@ -186,54 +200,67 @@ impl<T, S: Get<u32>> BoundedVec<T, S> {
 		S::get() as usize
 	}
 
-	/// Forces the insertion of `s` into `self` retaining all items with index at least `index`.
+	/// Forces the insertion of `element` into `self` retaining all items with index at least
+	/// `index`.
 	///
 	/// If `index == 0` and `self.len() == Self::bound()`, then this is a no-op.
 	///
 	/// If `Self::bound() < index` or `self.len() < index`, then this is also a no-op.
 	///
-	/// Returns `true` if the item was inserted.
-	pub fn force_insert_keep_right(&mut self, index: usize, element: T) -> bool {
+	/// Returns `true` if the item was inserted, and `Some(removed)` if an item was removed. By
+	/// definition, this can never return `(false, Some(_))`.
+	pub fn force_insert_keep_right(&mut self, index: usize, element: T) -> (bool, Option<T>)
+	where
+		T: Clone,
+	{
 		// Check against panics.
 		if Self::bound() < index || self.len() < index {
-			return false
-		}
-		if self.len() < Self::bound() {
+			(false, None)
+		} else if self.len() < Self::bound() {
 			// Cannot panic since self.len() >= index;
 			self.0.insert(index, element);
+			(true, None)
 		} else {
 			if index == 0 {
-				return false
+				return (false, None)
 			}
+			let removed = self[0].clone();
 			self[0] = element;
 			// `[0..index] cannot panic since self.len() >= index.
 			// `rotate_left(1)` cannot panic because there is at least 1 element.
 			self[0..index].rotate_left(1);
+			(true, Some(removed))
 		}
-		true
 	}
 
-	/// Forces the insertion of `s` into `self` retaining all items with index at most `index`.
+	/// Forces the insertion of `element` into `self` retaining all items with index at most
+	/// `index`.
 	///
 	/// If `index == Self::bound()` and `self.len() == Self::bound()`, then this is a no-op.
 	///
 	/// If `Self::bound() < index` or `self.len() < index`, then this is also a no-op.
 	///
-	/// Returns `true` if the item was inserted.
-	pub fn force_insert_keep_left(&mut self, index: usize, element: T) -> bool {
+	/// Returns `true` if the item was inserted, and `Some(removed)` if an item was been removed. By
+	/// definition, this can never return `(false, Some(_))`.
+	pub fn force_insert_keep_left(&mut self, index: usize, element: T) -> (bool, Option<T>)
+	where
+		T: Clone,
+	{
 		// Check against panics.
 		if Self::bound() < index || self.len() < index || Self::bound() == 0 {
-			return false
+			return (false, None)
 		}
 		// Noop condition.
 		if Self::bound() == index && self.len() <= Self::bound() {
-			return false
+			return (false, None)
 		}
+		// if we truncate anything, it will be this one.
+		let maybe_removed = self.0.get(Self::bound() - 1).cloned();
 		// Cannot panic since `Self.bound() > 0`
 		self.0.truncate(Self::bound() - 1);
```

Comment:
seems correct but since we need clone in the other function, I rather keep it like that (unless if you have a trick for the other case in your sleeve as well).